### PR TITLE
Fixes #3506: Image height remains zero for some time on bad network

### LINKED ISSUE #3506: Image View Height remains zero for some time on a bad network
**Image View height stays at zero for some time on bad network** 

This happens only sometimes but the image height in media detail remains zero for some time on a bad network then suddenly jumps into existence

It happens only for some images and only rarely

**i managed to capture it on a bad LTE network**
GIF-200313_125411 1

**System logs:**
Didn't find anything helpful 

**Would you like to work on the issue?**
Yes

### PR DESCRIPTION
**Pull request by kbhardwaj123**
**Optimised Width loading**

Fixes #3506 Image View Height remains zero for some time on a bad network

What changes did you make and why?
The calculation of image height used the width of scroll view which took time to inflate, now the width is retrieved from context which has almost no latency

### CODE REVIEW

File: app/src/main/java/fr/free/nrw/commons/media/MediaDetailFragment.java [hunk 1]
```diff
@@ -243,13 +246,22 @@ private void displayMediaDetails() {
...
             params.height = finalHeight;
             image.setLayoutParams(params);
         }
     }
 
+    private int getScreenWidth(Context context) {
+        DisplayMetrics dm = new DisplayMetrics();
+        WindowManager windowManager = (WindowManager) context.getSystemService(WINDOW_SERVICE);
+        windowManager.getDefaultDisplay().getMetrics(dm);
+        int width = dm.widthPixels;
```
Comment (macgills) [inline app/src/main/java/fr/free/nrw/commons/media/MediaDetailFragment.java]: you don't need to create a variable just `return dm.widthPixels`

**Comment by macgills** [comment]
You might consider one of the layoutListener's available to us, [this article](https://antonioleiva.com/kotlin-ongloballayoutlistener/) is mostly in kotlin but I think would better express the intention of this code

**Comment by neslihanturan** [comment]
I think it won't be possible merging #3505 before this PR is ready. Since this is a major issue. Can you please take a look at @macgills 's reviews @kbhardwaj123 ?

**Comment by kbhardwaj123** [comment]
@neslihanturan sure, I don't know somehow i didn't get notified about @macgills review on this will fix it ASAP

**Comment by kbhardwaj123** [comment]
@macgills i studied about the on global layout listener but i need you help on one thing, if we add a listener which returns the width after the layout is inflated then it might take some time and that itself might lead to latency, i mean i only need the width of the screen and not of the layout so why wait for the layout to get inflated and then for a listener to return it's width when we can directly have the width of the screen. I'm not sure though kindly correct me if i am wrong :)

**Comment by macgills** [comment]
@kbhardwaj123 getting the screen width means you are restricting this fragment to only ever be used in fullscreen display, in the android world often you have to consider many form factors which range from watched to tvs and god knows what next. If we displayed this on a tablet would it continue to work(after changing to a master/slave screen setup)? If we used multi window mode would it still work?

At the end of the day the screen is not ours, only our view hierarchy. I could be being overly cautious but I think it is good practice

**Comment by kbhardwaj123** [comment]
@macgills yes you are right i will see how to implement this here, thanks :)

**Comment by kbhardwaj123** [comment]
@macgills i have implemented the changes, now width will be taken from scrollview just as before, but in order to ensure that the getWidth is called after scrollView is drawn out i have used a view post runnable, refer: https://stackoverflow.com/questions/3591784/views-getwidth-and-getheight-returns-0

**Comment by macgills** [comment]
@kbhardwaj123 using ViewTreeObserver is a better expression of what we want to be doing here, while `post` does work I would prefer we were saying what we were doing even though it does look a bit more ugly. In kotlin we have `.doOnLayout` from [core-ktx](https://developer.android.com/kotlin/ktx) - quite surprising it has not been added - let me review the guidelines on adding libraries and hopefully it will be no problem and we can use that function, even though it will look a bit uglier from java it will still describe our intention

File: app/src/main/java/fr/free/nrw/commons/media/MediaDetailFragment.java [hunk 2]
```diff
@@ -9,11 +9,14 @@
 import android.os.Bundle;
 import android.text.Editable;
 import android.text.TextWatcher;
+import android.util.DisplayMetrics;
 import android.util.TypedValue;
 import android.view.LayoutInflater;
 import android.view.View;
 import android.view.ViewGroup;
 import android.view.ViewTreeObserver;
+import android.view.ViewTreeObserver.OnGlobalLayoutListener;
```
Comment (macgills) [inline app/src/main/java/fr/free/nrw/commons/media/MediaDetailFragment.java:18]: This should be the only new import I think

Comment (kbhardwaj123) [reply app/src/main/java/fr/free/nrw/commons/media/MediaDetailFragment.java:18]: yes indeed

Comment (kbhardwaj123) [reply app/src/main/java/fr/free/nrw/commons/media/MediaDetailFragment.java:18]: no it shows error otherwise, both imports would be required

Comment (macgills) [reply app/src/main/java/fr/free/nrw/commons/media/MediaDetailFragment.java:18]: image

The other imports, press `ctrl+alt+o` and turn these options on

image

Comment (kbhardwaj123) [reply app/src/main/java/fr/free/nrw/commons/media/MediaDetailFragment.java:18]: Oh I'll fix those asap

**Pull request merged**